common/rocksdb_env_s3.cpp

```diff
@@ -0,0 +1,432 @@
+    return Status::OK();
+  }
+};
+
+
+Status S3Env::NewWritableFile(const std::string& fname,
+                              std::unique_ptr<WritableFile>* result,
+                              const EnvOptions& options) {
+  assert(s3_util_ != nullptr);
+  result->reset();
+
+  std::unique_ptr<S3WritableFile> f(
+      new S3WritableFile(local_directory_ + fname, s3_key_prefix_ + fname, options, this, s3_util_));
+  auto s = f->status();
+  if (!s.ok()) {
+    LOG(ERROR) << "Error happened when creating S3WritableFile: " << s.ToString();;
+    return s;
+  }
+
+  result->reset(dynamic_cast<WritableFile*>(f.release()));
+  return Status::OK();
+}
+
+class S3Directory : public Directory {
+ public:
+  explicit S3Directory(S3Env* env, const std::string& name) : env_(env){
+    status_ = env_->GetBaseEnv()->NewDirectory(name, &posix_dir_);
+  }
+
+  ~S3Directory() {}
+
+  virtual Status Fsync() {
+    if (!status_.ok()) {
+      return status_;
+    }
+    return posix_dir_->Fsync();
+  }
+
+  virtual Status status() { return status_; }
+
+ private:
+  S3Env* env_;
+  Status status_;
+  std::unique_ptr<Directory> posix_dir_;
+};
+
+// S3 has no concepts of directories, so we just have to forward the request to
+// posix_env_ under the S3Directory
+Status S3Env::NewDirectory(const std::string& name,
+                           std::unique_ptr<Directory>* result) {
+  result->reset(nullptr);
+
+  // create new object.
+  std::unique_ptr<S3Directory> d(new S3Directory(this, name));
+
+  // Check if the path exists in local dir
+  if (!d->status().ok()) {
+    LOG(ERROR) << "Unable to create local dir: " << name;
+    return d->status();
+  }
+
+  result->reset(d.release());
+  return Status::OK();
+}
+
+Status S3Env::FileExists(const std::string& fname) {
+  // We read first from local storage and then from S3
+  auto st = posix_env_->FileExists(fname);
+  if (st.IsNotFound()) {
+    auto resp = s3_util_->getObjectMetadata(s3_key_prefix_ + fname);
+    if (!resp.Error().empty()) {
+      LOG(ERROR) << "Error happened when checking if file exists in S3: "
+                 << resp.Error();
+      return Status::NotFound();
+    }
+  }
+
+  return Status::OK();;
+}
+
+Status S3Env::GetChildren(const std::string& path,
+                          std::vector<std::string>* result) {
+  auto s3_full_path = s3_key_prefix_ + path;
+  // first fetch all files using the given path as the key prefix in S3
+  auto resp = s3_util_->listObjects(s3_full_path);
+  if (!resp.Error().empty()) {
+    LOG(ERROR) << "Error happened when fetching files from S3: "
+               << resp.Error() << " under path: " << s3_full_path;
+    return Status::IOError();
+  }
+
+  // trim the file name
+  for (auto& v : resp.Body()) {
+    // Our path should be a prefix of the fetched value
+    if (v.find(s3_full_path) == 0) {
+      result->push_back(v.substr(s3_full_path.size()));
+    }
+  }
+
+  // then fetch all files that exist in the local posix directory
```

Comment:
If you have already read from s3, what is the purpose of reading from local dir? Is it for cases when changes have been made locally but the file has not been closed?

Reply:
yes. we wont need to read them from local again. I already fixed it